Test/tutorial/parser_combinators.dfy

```diff
@@ -0,0 +1,265 @@
+/// an argument, without having to pass it around everywhere and without having
+/// to specify that multiple calls use the same value of it: we define a
+/// `Engine` to be a datatype that contains the string, and all members of the
+/// datatype can access it (though note that most of them don't).
+
+  datatype Engine = Engine(input: string) {
+
+/// `Concat` takes two parsers and returns a lambda:
+
+    static function Concat0<L, R>(left: Parser<L>, right: Parser<R>)
+      : (p: Parser<(L, R)>)
+    {
+      (pos: nat)
+        requires left.requires(pos)
+        requires left(pos).Success? ==> right.requires(left(pos).pos)
+      =>
+        var (pos, l) :- left(pos);
+        var (pos, r) :- right(pos);
+        Success(pos, (l, r))
+    }
+
+/// The definition above is transparent: every time the verifier encounters it,
+/// it is allowed to peek inside the body of the function and use it for
+/// reasoning.  This will cause verification-performance issues down the line,
+/// so we use the following alternative definition instead:
+
+    static function {:opaque} Concat<L, R>(
+      left: Parser<L>,
+      right: Parser<R>
+    )
+      : (p: Parser<(L, R)>)
+      ensures forall pos: nat |
+        && left.requires(pos)
+        && (left(pos).Success? ==> right.requires(left(pos).pos))
+        :: p.requires(pos)
+    {
+      (pos: nat)
+        requires left.requires(pos)
+        requires left(pos).Success? ==> right.requires(left(pos).pos)
+      =>
+        var (pos, l) :- left(pos);
+        var (pos, r) :- right(pos);
+        Success(pos, (l, r))
+    }
+
+/// The two important things to note are (1) the fact that `Concat` is now
+/// marked `:opaque`, and (2) the fact that `Concat` now has an `ensures` clause
+/// of the following form:
+///
+/// ```dafny
+///   ensures forall input :: (precondition of the parser) ==> p.requires(input)
+/// ```
+///
+/// This does not say that the parser must accept all inputs: only that the
+/// condition `(precondition of the parser)` is sufficient to call the opaque
+/// function `p` that `Concat` returns.
+///
+/// With that in mind, we can define `Either` in the same way:
+
+    static function {:opaque} Either<L, R>(
+      left: Parser<L>,
+      right: Parser<R>
+    )
+      : (p: Parser<Either<L, R>>)
+      ensures forall pos: nat |
+        && left.requires(pos)
+        && (left(pos).Failure? ==> right.requires(pos))
+        :: p.requires(pos)
+    {
+      (pos: nat)
+        requires left.requires(pos)
+        requires left(pos).Failure? ==> right.requires(pos)
+      =>
+        match left(pos)
+          case Failure => right(pos).MapResult(r => Right(r))
+          case Success(pos, l) => Success(pos, Left(l))
+    }
+
+/// Next, we define `Char`, which checks for a specific character and advances
+/// the cursor by one.  Its post-condition indicates that the cursor has
+/// advanced and that there was some input left.
```

Comment:
"was" -> "is"

Reply:
`was` was right, but I have now changed the postcondition to avoid this confusing statement.